src/java.base/share/classes/java/io/Console.java

```diff
@@ -594,23 +597,37 @@ public int read(char[] cbuf, int offset, int length)
 
         // Set up JavaIOAccess in SharedSecrets
         SharedSecrets.setJavaIOAccess(new JavaIOAccess() {
+            @SuppressWarnings("removal")
             public Console console() {
-                if (istty) {
-                    if (cons == null)
-                        cons = new Console();
-                    return cons;
+                if (cons == null) {
+                    try {
+                        // Try loading providers
+                        PrivilegedAction<Console> pa = () -> {
+                            var consModName = System.getProperty("jdk.console",
+                                    JdkConsoleProvider.DEFAULT_PROVIDER);
+                            return ServiceLoader.load(JdkConsoleProvider.class).stream()
+                                    .map(ServiceLoader.Provider::get)
+                                    .filter(jcp -> consModName.equals(jcp.getClass().getModule().getName()))
+                                    .map(jcp -> jcp.console(istty))
+                                    .filter(Objects::nonNull)
+                                    .findAny()
+                                    .map(jc -> (Console) new ProxyingConsole(jc))
+                                    .orElse(istty ? new Console() : null);
+                        };
+                        cons = AccessController.doPrivileged(pa);
+                    } catch (ServiceConfigurationError ignore) {
+                        // default to built-in Console
+                        cons = istty ? new Console() : null;
+                    }
                 }
-                return null;
-            }
-
-            public Charset charset() {
-                return CHARSET;
+                return cons;
             }
         });
     }
     private static Console cons;
```

Comment:
The initialization of cons is replaced in this PR so maybe we can make it final at the same time.

Reply:
Changed it to final